src/controllers/hid/hidiothread.cpp

```diff
@@ -0,0 +1,301 @@
+    for (unsigned char i = 0; i < m_outputReports.size(); i++) {
+        {
+            auto lock = lockMutex(&m_outputReportMapMutex);
+            m_outputReportIterator++;
+            if (m_outputReportIterator == m_outputReports.end()) {
+                m_outputReportIterator = m_outputReports.begin();
+            }
+        }
+
+        // The only mutable operation on m_outputReports is insert
+        // by std::map<Key,T,Compare,Allocator>::operator[]
+        // The standard says that "No iterators or references are invalidated." using this operator.
+        // Therefore m_outputReportIterator doesn't require Mutex protection.
+        if (m_outputReportIterator->second->sendOutputReport(
+                    &m_hidDeviceMutex, m_pHidDevice, m_deviceInfo, m_logOutput)) {
+            // Return after each time consuming sendOutputReport
+            return true;
+        }
+    }
+    // Returns false if no report required a time consuming sendOutputReport
+    return false;
+}
+
+void HidIoThread::sendFeatureReport(
+        const QByteArray& reportData, unsigned int reportID) {
+    auto startOfHidSendFeatureReport = mixxx::Time::elapsed();
+    QByteArray dataArray;
+    dataArray.reserve(kReportIdSize + reportData.size());
+
+    // Append the Report ID to the beginning of dataArray[] per the API..
+    dataArray.append(reportID);
+
+    for (const int datum : reportData) {
+        dataArray.append(datum);
+    }
+
+    auto lock = lockMutex(&m_hidDeviceMutex);
+    int result = hid_send_feature_report(m_pHidDevice,
+            reinterpret_cast<const unsigned char*>(dataArray.constData()),
+            dataArray.size());
+    if (result == -1) {
+        qCWarning(m_logOutput)
+                << "sendFeatureReport is unable to send data to"
+                << m_deviceInfo.formatName() << "serial #"
+                << m_deviceInfo.serialNumber() << ":"
+                << mixxx::convertWCStringToQString(
+                           hid_error(m_pHidDevice), kMaxHidErrorMessageSize);
+        return;
+    }
+
+    qCDebug(m_logOutput)
+            << result << "bytes sent by sendFeatureReport to"
+            << m_deviceInfo.formatName() << "serial #"
+            << m_deviceInfo.serialNumber() << "(including report ID of"
+            << reportID << ") - Needed: "
+            << (mixxx::Time::elapsed() - startOfHidSendFeatureReport)
+                       .formatMicrosWithUnit();
+}
+
+QByteArray HidIoThread::getFeatureReport(
+        unsigned int reportID) {
+    auto startOfHidGetFeatureReport = mixxx::Time::elapsed();
+    unsigned char dataRead[kReportIdSize + kBufferSize];
+    dataRead[0] = reportID;
+
+    auto lock = lockMutex(&m_hidDeviceMutex);
+    int bytesRead = hid_get_feature_report(m_pHidDevice,
+            dataRead,
+            kReportIdSize + kBufferSize);
+    if (bytesRead <= kReportIdSize) {
+        // -1 is the only error value according to hidapi documentation.
```

Comment:
Obtain the string with hid_error() and release the lock before logging a message. I guess mixxx::convertWCStringToQString() must be invoked while still locked. Store the result in a local variable and then unlock().

Reply:
The `qCWarning `messages, which use hid_error occur extremly seldom. Usually they only occur if your mapping is wrong (e.g. Report lenght is defined wrong) or you disconnect the device in operation. I would prefer code simplicity over performance in these cases.
Different are the `qCDebug `messages which always occur if `--controllerDebug` is specified. Here performance matters and they shouldn't block other operation.